intellij/src/main/kotlin/motif/intellij/MotifProjectComponent.kt

```diff
@@ -0,0 +1,152 @@
+                override fun run(indicator: ProgressIndicator) {
+                    ApplicationManager.getApplication().runReadAction {
+                        val graph: ResolvedGraph = graphFactory.compute()
+                        onGraphUpdated(graph)
+                    }
+                }
+            })
+        }
+    }
+
+    fun refreshGraph() {
+        if (isRefreshing) {
+            return
+        }
+        isRefreshing = true
+
+        ProgressManager.getInstance().run(object : Task.Backgroundable(project, LABEL_GRAPH_REFRESH) {
+            override fun run(indicator: ProgressIndicator) {
+                ApplicationManager.getApplication().runReadAction {
+                    val updateGraph: ResolvedGraph = graphFactory.compute()
+                    onGraphUpdated(updateGraph)
+                    isRefreshing = false
+                }
+            }
+        })
+    }
+
+    fun onSelectedScope(element: PsiElement) {
+        if (element !is PsiClass || !isMotifScopeClass(element)) {
+            return
+        }
+        scopePanel?.setSelectedScope(element)
+    }
+
+    private fun onGraphUpdated(graph: ResolvedGraph) {
+        ApplicationManager.getApplication().invokeLater {
+            val toolWindowManager: ToolWindowManager = ToolWindowManager.getInstance(project)
+            if (toolWindowManager.getToolWindow(TOOL_WINDOW_ID) == null) {
+                val toolWindow: ToolWindow = toolWindowManager.registerToolWindow(TOOL_WINDOW_ID, true, ToolWindowAnchor.RIGHT)
+                toolWindow.title = TOOL_WINDOW_TITLE
+
+                scopePanel = MotifScopePanel(project, graph)
+                val scopesContent: Content = ContentFactory.SERVICE.getInstance().createContent(scopePanel, TAB_NAME_SCOPES, true)
+                scopesContent.isCloseable = false
+                toolWindow.contentManager.addContent(scopesContent)
+
+                error = MotifErrorPanel(project, graph)
+            } else {
+                scopePanel?.onGraphUpdated(graph)
+                error?.onGraphUpdated(graph)
+            }
+
+            // Update the visibility of the error tab
+            updateErrorTabVisibility(graph)
+
+            // Propagate changes to line markers provider
+            val language: Language? = Language.findLanguageByID("JAVA")
+            if (language != null) {
+                for (lineMarkerProvider in LineMarkerProviders.INSTANCE.allForLanguage(language)) {
+                    if (lineMarkerProvider is ScopeNavigationLineMarkerProvider) {
+                        lineMarkerProvider.onGraphUpdated(graph)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun updateErrorTabVisibility(graph: ResolvedGraph) {
+        val toolWindow: ToolWindow = ToolWindowManager.getInstance(project).getToolWindow(TOOL_WINDOW_ID) ?: return
+        if (graph.errors.isEmpty()) {
+            if (errorContent != null) {
```

Comment:
For executing code only if something is non-null use: `foo?.let { <"it" is non-null in here> }`. See [Safe Calls](https://kotlinlang.org/docs/reference/null-safety.html#safe-calls)
```kotlin
errorContent?.let { toolWindow.contentManager.removeContent(it, true) }
```